### PR TITLE
Refs [TASK][44597] following coins screen

### PR DESCRIPTION
**Pull request by ghost**
Related Tickets
#Task 44597: Content
WHAT
Evidence (Screenshot or Video)
image

Review Checklist
Category	View Point	Description	Self review	Reviewer2 (name)
Conventions	Sun* coding conventions	https://github.com/framgia/coding-standards/tree/master/eng/android	
 yes
 yes
Conventions	Kotlin coding conventions	https://kotlinlang.org/docs/coding-conventions.html	
 yes
 yes
Redmine	Sun* Redmine working process	https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md	
 yes
 yes
Notes (Optional)
(Impacted Areas in Application(List features, api, models or services that this PR will affect))

(List gem, library third party add new)

(Other notes)

### CODE REVIEW

File: lib/bloc/list_coin/list_coin_bloc.dart [hunk 1]
```diff
@@ -1,26 +1,54 @@
...
+      for (var followingCoin in followingCoins) {
+        for (var coin in coins) {
+          if (followingCoin.coinId == coin.id) {
+            coin.isFollowing = true;
+            break;
+          }
+        }
+      }
+    }
+
```
Comment (tuannm-1276) [inline lib/bloc/list_coin/list_coin_bloc.dart]: có cách nào viết xịn hơn ko em :D

File: lib/bloc/list_coin/list_coin_event.dart [hunk 2]
```diff
@@ -1,8 +1,21 @@
...
+  final CoinLocal coin;
+
+  FollowingCoin(this.coin);
+}
+
+class UnFollowingCoin extends ListCoinEvent {
+  final String coinId;
+
+  UnFollowingCoin(this.coinId);
+}
```
Comment (tuannm-1276) [inline lib/bloc/list_coin/list_coin_event.dart]: bl

File: lib/data/model/item_coin.dart [hunk 3]
```diff
@@ -23,9 +24,23 @@ class ItemCoin {
...
+          coin.symbol,
+          coin.marketData.currentPrice.usd,
+          coin.marketData.marketCap.usd,
+          coin.marketCapRank,
+          coin.marketData.priceChangePercentage24h,
+          coin.image.large);
+
   Map<String, dynamic> toJson() => _$ItemCoinToJson(this);
+
+
```
Comment (tuannm-1276) [inline lib/data/model/item_coin.dart]: bl

File: lib/data/repository/follow_repository.dart [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+class FollowRepository {
+  DatabaseProvider databaseProvider = getIt<DatabaseProvider>();
+
+  Future<List<CoinLocal>> getFollowingCoins() => databaseProvider.getCoins();
+
+  Future<int> addFollowing(CoinLocal coin) => databaseProvider.insertCoin(coin);
+
+  Future<int> deleteFollowing(String coinId) => databaseProvider.deleteCoin(coinId);
+}
```
Comment (tuannm-1276) [inline lib/data/repository/follow_repository.dart]: bl

File: lib/data/database/database.dart [hunk 5]
```diff
@@ -25,14 +24,37 @@ class DatabaseProvider {
...
+  }
+  
+  Future<int> deleteCoin(String coinId) async {
+    final db = await database;
+    return await db.delete(FOLLOWING_TABLE, where: "id = ?", whereArgs: [coinId]);
+  }
+
   Future close() {
-    return _database.close();
+    return _database!.close();
```
Comment (tuannm-1276) [inline lib/data/database/database.dart:57]:
```suggestion
    return _database?.close();
```

File: lib/data/database/database.dart [hunk 6]
```diff
@@ -1,22 +1,21 @@
...
 import 'package:sqflite/sqflite.dart';
+import 'package:vn_crypto/data/model/coin_local.dart';
 
 const String DB_NAME = "vncrypto_db";
 const String FOLLOWING_TABLE = "following";
 
 class DatabaseProvider {
   static final DatabaseProvider databaseProvider = DatabaseProvider.init();
-  late Database _database;
+  static Database? _database;
```
Comment (tuannm-1276) [inline lib/data/database/database.dart]:
```suggestion
  Database _database;
```

File: lib/ui/screen/CoinDetailsScreen.dart [hunk 7]
```diff
@@ -6,34 +6,50 @@ import 'package:url_launcher/link.dart';
...
+import 'package:vn_crypto/data/model/item_coin.dart';
 import 'package:vn_crypto/data/repository/coins_repository.dart';
+import 'package:vn_crypto/data/repository/follow_repository.dart';
 import 'package:vn_crypto/di/dependency_injection.dart';
 import 'package:vn_crypto/ui/components/common/price_change_with_border.dart';
 import 'package:vn_crypto/ultils/Constant.dart';
 
-class CoinDetailsScreen extends StatelessWidget {
-  final String coinId;
+class CoinDetailsScreen extends StatefulWidget {
```
Comment (tuannm-1276) [inline lib/ui/screen/CoinDetailsScreen.dart:18]: view này to quá, e tách ra các view con nhé

File: lib/data/database/database.dart (hunk 5 above)
Comment (ghost) [reply lib/data/database/database.dart:57]: bên flutter không có ? anh ạ, có mỗi ! thôi :(

**Review by vietnh-1109 — APPROVED** [review]

**Review by tuannm-1276 — APPROVED** [review]

**Pull request merged**